bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/handler/EcovacsVacuumHandler.java

```diff
@@ -0,0 +1,830 @@
+                        roomIds.add(String.valueOf(id.charAt(0) - 'A'));
+                    } else {
+                        logger.info("{}: Found invalid spot area room ID '{}', ignoring.", getDeviceSerial(), id);
+                    }
+                }
+                if (!roomIds.isEmpty()) {
+                    return new SpotAreaCleaningCommand(roomIds, passes);
+                }
+            } else {
+                logger.info("{}: spotArea command needs to have the form spotArea:<room1>[;<room2>][;<...roomX>][:x2]",
+                        getDeviceSerial());
+            }
+        }
+        if (command.startsWith(CMD_CUSTOM_AREA) && device.hasCapability(DeviceCapability.CUSTOM_AREA_CLEANING)) {
+            String[] splitted = command.split(":");
+            if (splitted.length == 2 || splitted.length == 3) {
+                String coords = splitted[1];
+                int passes = splitted.length == 3 && "x2".equals(splitted[2]) ? 2 : lastDefaultCleaningPasses;
+                String[] splittedAreaDef = coords.split(";");
+                if (splittedAreaDef.length == 4) {
+                    return new CustomAreaCleaningCommand(String.join(",", splittedAreaDef), passes);
+                }
+            }
+            logger.info("{}: customArea command needs to have the form customArea:<x1>;<y1>;<x2>;<y2>[:x2]",
+                    getDeviceSerial());
+        }
+
+        return null;
+    }
+
+    private interface WithDeviceAction {
+        void run(EcovacsDevice device) throws EcovacsApiException, InterruptedException;
+    }
+
+    private void doWithDevice(WithDeviceAction action) {
+        EcovacsDevice device = this.device;
+        if (device == null) {
+            return;
+        }
+        try {
+            action.run(device);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } catch (EcovacsApiException e) {
+            logger.debug("{}: Failed communicating to device, reconnecting", getDeviceSerial(), e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            if (e.isAuthFailure) {
+                EcovacsApiHandler apiHandler = getApiHandler();
+                if (apiHandler != null) {
+                    apiHandler.onLoginExpired();
+                }
+                // Drop our device instance to make sure we run a full init cycle,
+                // including an API re-login, on reconnection
+                device.disconnect(scheduler);
+                this.device = null;
+            }
+            teardownAndScheduleReconnection();
+        }
+    }
+
+    private String getDeviceSerial() {
+        final String serialProp = getThing().getProperties().getOrDefault(Thing.PROPERTY_SERIAL_NUMBER, "");
+        return serialProp.isEmpty() ? getConfigAs(EcovacsVacuumConfiguration.class).serialNumber : serialProp;
```

Comment:
Serial number seems required in configuration, is there any reason to look in properties? In which case will there be a valid serial number property, but not any configured one for the thing?

Reply:
The thought was that on automatic discovery, properties are filled from the `DiscoveryResult`, but configuration might not be. After re-reading the docs, one can read it as if configuration is always pre-filled from the `DiscoveryResult`. Is this guaranteed to be the case?

Reply:
> After re-reading the docs, one can read it as if configuration is always pre-filled from the `DiscoveryResult`. Is this guaranteed to be the case?

Correct, properties set by discovery will populate initial configuration, so you shouldn't need to use the properties in your logic.

> - for discovered things, what's used is always the discovered serial number
> - for manually added things, what's used is always the configured serial number (as no serial property is set in that case)

Third option: Manually configured things in files: no serial number.

Does it matter that serial number might be included in the Thing UID? The same is true for other bindings, for example the Miele binding:

https://github.com/openhab/openhab-addons/blob/dae33c405ee482d1cb226dfe4de37a57ba775ecf/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleApplianceDiscoveryService.java#L155

And the configuration is also populated with appliance id (which is serial number):

https://github.com/openhab/openhab-addons/blob/dae33c405ee482d1cb226dfe4de37a57ba775ecf/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleApplianceDiscoveryService.java#L113

And used:

https://github.com/openhab/openhab-addons/blob/dae33c405ee482d1cb226dfe4de37a57ba775ecf/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleApplianceHandler.java#L147

Reply:
> Does it matter that serial number might be included in the Thing UID?

That's the question ;-) I guess if a user is to do that edit, they hopefully know what they're doing. As I see the Miele binding isn't doing dynamic updating (as in: `handleConfigurationUpdate` handling) for the serial number either, I guess this should be OK in my case as well. I updated the handling accordingly.